CI: Install KBS on k8s for attestation tests

wainersm opened the pull request:
## Introduction

This introduces the building blocks to install the [CoCo KBS](https://github.com/confidential-containers/kbs) on Kubernetes to be used on the upcoming attestation tests (see https://github.com/kata-containers/kata-containers/issues/9058).

It is added the commands `deploy-coco-kbs` and `delete-coco-kbs` to the `tests/integration/kubernetes/gha.sh` script to install and uninstall the KBS on Kubernetes.

I tested those new commands on a kubeadm cluster as:

```
$ ./tests/integration/kubernetes/gha-run.sh deploy-coco-kbs
::group::Clone the kbs sources
Cloning into '/tmp/kbs'...
remote: Enumerating objects: 283, done.
remote: Counting objects: 100% (283/283), done.
remote: Compressing objects: 100% (256/256), done.
remote: Total 283 (delta 24), reused 185 (delta 8), pack-reused 0
Receiving objects: 100% (283/283), 261.03 KiB | 1.16 MiB/s, done.
Resolving deltas: 100% (24/24), done.
/tmp/kbs ~/go/src/github.com/kata-containers/kata-containers
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
From https://github.com/confidential-containers/kbs
 * branch            112018371d38c73eba8998f4f6b81ebe00b41556 -> FETCH_HEAD
Switched to a new branch 'kbs_126625'
::endgroup::
/tmp/kbs/kbs/config/kubernetes /tmp/kbs ~/go/src/github.com/kata-containers/kata-containers
::group::Update the kbs container image
/tmp/kbs/kbs/config/kubernetes/base /tmp/kbs/kbs/config/kubernetes /tmp/kbs ~/go/src/github.com/kata-containers/kata-containers
/tmp/kbs/kbs/config/kubernetes /tmp/kbs ~/go/src/github.com/kata-containers/kata-containers
::endgroup::
::group::Deploy the KBS
namespace/coco-tenant created
configmap/kbs-config-k5gm89bk2b created
secret/kbs-auth-public-key-26bb2m572t created
secret/keys-gmf2g75547 created
service/kbs created
deployment.apps/kbs created
/tmp/kbs ~/go/src/github.com/kata-containers/kata-containers
~/go/src/github.com/kata-containers/kata-containers
pod/kbs-598c4bfd-qrs26 condition met
::endgroup::
::group::Check the service healthy
pod/kbs-checker-126625 created
wget: server returned error: HTTP/1.1 404 Not Found
KBS service respond to requests
::endgroup::
``` 

And to delete:

```
$ ./tests/integration/kubernetes/gha-run.sh delete-coco-kbs
/tmp/kbs ~/go/src/github.com/kata-containers/kata-containers
namespace "coco-tenant" deleted
configmap "kbs-config-k5gm89bk2b" deleted
secret "kbs-auth-public-key-26bb2m572t" deleted
secret "keys-gmf2g75547" deleted
service "kbs" deleted
deployment.apps "kbs" deleted
~/go/src/github.com/kata-containers/kata-containers
```

As agreed on the CoCo CI meeting we will pursued to run the attestation tests on both TEE and non-TEE jobs. Initially I've changed the non-TEE [run-kata-deploy-tests-on-aks.yaml](https://github.com/kata-containers/kata-containers/blob/main/.github/workflows/run-kata-deploy-tests-on-aks.yaml) workflow to get KBS installed, but only for the kata-qemu configuration. Further workflows should be addressed on follow up PRs.

Fixes #9056

## Notes

 * As mentioned the code was tested on a kubeadm (not exposing the service) and AKS clusters, ~I hope to run the CI to see if it works for AKS (in theory it should)~. For other cluster types, e.g., k3s, we will need to test it too but hopefully only when it comes the time to run attestation tests on workflows that use those "alternative" k8s clusters
 * ~Currently the KBS service is reachable only within the cluster (as one can note on the "healthy checker")~. Depending on how we gonna feed the KBS with data for the tests, we might need to expose the service with a public IP. For example, suppose that we want to use the [KBS client](https://github.com/confidential-containers/kbs/tree/main/kbs/tools/client) from within the Github Actions runner, then it will be required to expose the service externally from AKS. Other questions will derive from this, for example, how/where to install the client, how to get access to the authentication key...etc @GabyCT @fitzthum something we will need to sync'up
 * The first commit of this series, that changes the workflow, should be merged before and in another PR. I leaving it here to receive the initial reviews as well as show how it all integrate

fitzthum commented:
This seems great. Wainer mentioned a couple of discussion points which are still pending.

For the private key, can we just expect the test to get it from the filesystem? The location is known.

fitzthum commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -0,0 +1,109 @@
+
+	# Tests should fill kbs resources later, however, the deployment
+	# expects at least one secret served at install time.
+	echo "somesecret" > overlays/key.bin
```
Ok. I guess we could change this script upstream to note require any secrets, but it's not a big deal.

jodh-intel commented on .github/workflows/run-k8s-tests-on-aks.yaml:
```diff
@@ -103,7 +103,12 @@ jobs:
-      
+
+      - name: Deploy CoCo KBS
+        if: ${{ matrix.vmm == 'qemu' }}
```
I wonder if it's worth passing `${{matrix.vmm}}` to the `gha-run.sh` script below so we can expand it to support other hypervisors in future?

wainersm replied:
`${{matrix.vmm}}` is exported in `KATA_HYPERVISOR` that I can read like from `gha-run.sh`

Are you thinking in have a filter similar to https://github.com/kata-containers/kata-containers/pull/9062/commits/4daed2292bb49b7c3d8b301de0ff7f02354e2f9b ?

sprt replied:
@wainersm How about putting this condition in the shell script so upcoming PRs can easily modify it?

wainersm replied:
Done.

jodh-intel commented on tests/gha-run-k8s-common.sh:
```diff
@@ -94,6 +94,25 @@ function install_kubectl() {
+	arch=$(arch_to_golang)
+
+	curl -L "https://github.com/kubernetes-sigs/kustomize/releases/download/kustomize/${version}/kustomize_${version}_linux_${arch}.tar.gz" | \
+		sudo tar xvzf - -C /usr/local/bin
```
Nit: For consistency, I'd consider adding the `dash` to the other options:

```suggestion
		sudo tar -xvzf - -C /usr/local/bin
```

jodh-intel commented on tests/gha-run-k8s-common.sh:
```diff
@@ -94,6 +94,25 @@ function install_kubectl() {
+	arch=$(arch_to_golang)
+
+	curl -L "https://github.com/kubernetes-sigs/kustomize/releases/download/kustomize/${version}/kustomize_${version}_linux_${arch}.tar.gz" | \
+		sudo tar xvzf - -C /usr/local/bin
```
I think we should be paranoid here:

- Download the tar file to a `mktemp -d` directory.
- Download the checksum file (for example, https://github.com/kubernetes-sigs/kustomize/releases/download/kustomize%2Fv5.3.0/checksums.txt).
- Check the tar file against the checksum.
- Assuming the checksum is correct, unpack and then delete the temp directory.

GabyCT replied:
what about also implementing a trap to ensure that the temp directory is deleted
```
TEMP_DIR="$( mktemp -d )"
 trap 'rm -rf ${TEMP_DIR}' EXIT
```

jodh-intel commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -0,0 +1,109 @@
+source "${kubernetes_dir}/../../gha-run-k8s-common.sh"
+
+# Where the kbs sources will be cloned
+readonly COCO_KBS_DIR="/tmp/kbs"
```
Can you use `mktemp -d` here with a global variable and set it in a `setup()` function or similar.

wainersm replied:
hi @jodh-intel , I also like the idea of using a temp dir, my only concern on this case is the added complexity. The functions will be called on different GHA steps, I don't think exported variables on one will be carried out to the next. I can think in a couple of options though:
 * save $COCO_KBS_DIR as an step output
 * save $COCO_KBS_DIR in $GITHUB_ENV, which is used by github actions to export variables
 
However both options works only on github context; running locally we will need another logic.

Worth tackling this problem?

jodh-intel commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -0,0 +1,109 @@
+	image=$(get_from_kata_deps "externals.coco-kbs.image")
+	image_tag=$(get_from_kata_deps "externals.coco-kbs.image_tag")
+
+	if [ -d "$COCO_KBS_DIR" ]; then
+		rm -rf "$COCO_KBS_DIR"
+	fi
```
Personal preference, but you could collapse this to a single line:

```suggestion
	[ -d "$COCO_KBS_DIR" ] && rm -rf "$COCO_KBS_DIR"
```

GabyCT replied:
maybe also add a trap to see that is deleted

jodh-intel commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -0,0 +1,109 @@
+	git clone --depth 1 "${repo}" "$COCO_KBS_DIR"
+	pushd "$COCO_KBS_DIR"
+	git fetch --depth=1 origin "${version}"
+	git checkout FETCH_HEAD -b kbs_$$
```
Do we need the `PID` suffix if we make `COCO_KBS_DIR` unique?

fitzthum replied:
If we are going to store the kbs key here (and access it from the tests), we might not want this path to change.

wainersm replied:
Good point @jodh-intel. In theory it can be just `kbs` or something well; and hope no one has the great idea to name a remote branch like that :)

@fitzthum this is just the temporary branch name. We can even checkout FETCH_HEAD without branching.

ldoktor replied:
This checks the main and then re-fetches the desired version, how about using:

```bash
git init $COCO_KBS_DIR
git remote add origin https://github.com/confidential-containers/kbs
git fetch --depth 1 origin "${version}"
git checkout FETCH_HEAD
```

(also I don't see a reason to branch the FETCH_HEAD, it should be sufficient to check it out, shouldn't it?)

wainersm replied:
hi @ldoktor , check FETCH_HEAD out should be sufficient but I've this habit of branching to silence the `You are in 'detached HEAD' state` warning.

jodh-intel commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -0,0 +1,109 @@
+	fi
+
+	# Give some time for the service to start.
+	sleep 5
```
Is there any way to avoid this since it's bound to break somewhere.

wainersm replied:
Yes, I can think on a couple of solutions:

* Keep grep'ing the pod logs until we find a known message (or hit timeout)
* Implement some sort of liveness probe https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/
* Keep running the checker below until it gets a response (i.e. `404 Not Found`) (or hit timeout)

Any other suggestion?

GabyCT replied:
for me the first option is valid we can keep seeing the pod until we find a message for a period of time and then fail if it fails it

ldoktor replied:
The thing with grepping is that sometimes kubectl fails, ... . Perhaps we could introduce a new function to wait for a number of consequent passes...

jodh-intel commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -0,0 +1,109 @@
+	kbs_port=$(kubectl get -o jsonpath='{.spec.ports[0].port}' service/kbs -n "$ns" 2>/dev/null)
+	local pod=kbs-checker-$$
+	kubectl run "$pod" --image=busybox --restart=Never --wait=true -- \
+		sh -c "wget -O- \"${kbs_ip}:${kbs_port}\" || true"
```
We prefer `curl` over `wget`.

wainersm replied:
The busybox image doesn't come with `curl`

jodh-intel commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -0,0 +1,109 @@
+	kubectl run "$pod" --image=busybox --restart=Never --wait=true -- \
+		sh -c "wget -O- \"${kbs_ip}:${kbs_port}\" || true"
+	waitForProcess "30" "10" "kubectl get pods/${pod} | grep -q Completed"
+	if ! kubectl logs "$pod" | grep "404 Not Found"; then
```
```suggestion
	if ! kubectl logs "$pod" | grep -q "404 Not Found"; then
```

wainersm replied:
Changed on updated PR.

jodh-intel reviewed (commented):
Thanks, @wainersm - A few comments...

GabyCT commented on tests/gha-run-k8s-common.sh:
```diff
@@ -94,6 +94,25 @@ function install_kubectl() {
+	fi
+
+	ensure_yq
+	version=$(get_from_kata_deps "externals.kustomize.version")
```
It is not a big deal but I prefer put it the local here `local version` `local arch`

wainersm commented:
> Thanks, @wainersm - A few comments...


Hey @jodh-intel ! I'm happy to see your comments to my code again! I will be addressing them soon.

wainersm commented:
> This seems great. Wainer mentioned a couple of discussion points which are still pending.
> 
> For the private key, can we just expect the test to get it from the filesystem? The location is known.

Yes, I will add a function like `kbs_get_private_key()`

wainersm commented:
Updated the PR so that It isn't WIP anymore (CC @fidencio).

The way it is implemented, `./tests/integration/kubernetes/gha-run.sh deploy-coco-kbs` assume you are running on AKS. It will install the HTTP application routing add-on and configure Ingress so that the service is exposed publicly; it also check that both the service and the public endpoint respond to requests.

Tested installation as:

```shell
$ export AZ_RG="wmoschet-ci-tests"
$ export AKS_NAME="wmoschet-aks-ci-test"
$ cd tests
$ ./integration/kubernetes/gha-run.sh create-cluster
< omitted output>
$ ./integration/kubernetes/gha-run.sh get-cluster-credentials
< omitted output>
$ ./integration/kubernetes/gha-run.sh deploy-coco-kbs
::group::Clone the kbs sources
Cloning into '/tmp/kbs'...
remote: Enumerating objects: 282, done.
remote: Counting objects: 100% (282/282), done.
remote: Compressing objects: 100% (256/256), done.
remote: Total 282 (delta 24), reused 185 (delta 7), pack-reused 0
Receiving objects: 100% (282/282), 260.80 KiB | 1.38 MiB/s, done.
Resolving deltas: 100% (24/24), done.
/tmp/kbs ~/go/src/github.com/kata-containers/kata-containers/tests
remote: Enumerating objects: 23, done.
remote: Counting objects: 100% (23/23), done.
remote: Compressing objects: 100% (13/13), done.
remote: Total 13 (delta 10), reused 2 (delta 0), pack-reused 0
Unpacking objects: 100% (13/13), 1.77 KiB | 453.00 KiB/s, done.
From https://github.com/confidential-containers/kbs
 * branch            112018371d38c73eba8998f4f6b81ebe00b41556 -> FETCH_HEAD
Switched to a new branch 'kbs_300525'
::endgroup::
/tmp/kbs/kbs/config/kubernetes /tmp/kbs ~/go/src/github.com/kata-containers/kata-containers/tests
::group::Update the kbs container image
/tmp/kbs/kbs/config/kubernetes/base /tmp/kbs/kbs/config/kubernetes /tmp/kbs ~/go/src/github.com/kata-containers/kata-containers/tests
/tmp/kbs/kbs/config/kubernetes /tmp/kbs ~/go/src/github.com/kata-containers/kata-containers/tests
::endgroup::
::group::Enable HTTP application routing add-on
{
  "aadProfile": null,
  "addonProfiles": {
    "httpApplicationRouting": {
      "config": {
        "HTTPApplicationRoutingZoneName": "77b2a8bfe3c8468385d0.eastus.aksapp.io"
      },
      "enabled": true,

< omitted output >

  }
}
::endgroup::
/tmp/kbs/kbs/config/kubernetes/overlays /tmp/kbs/kbs/config/kubernetes /tmp/kbs ~/go/src/github.com/kata-containers/kata-containers/tests
::group::/tmp/kbs/kbs/config/kubernetes/overlays/ingress.yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: kbs
  namespace: coco-tenant
  annotations:
    kubernetes.io/ingress.class: addon-http-application-routing
spec:
  rules:
  - host: kbs.77b2a8bfe3c8468385d0.eastus.aksapp.io
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: kbs
            port:
              number: 8080
::endgroup::
/tmp/kbs/kbs/config/kubernetes /tmp/kbs ~/go/src/github.com/kata-containers/kata-containers/tests
::group::Deploy the KBS
namespace/coco-tenant created
configmap/kbs-config-k5gm89bk2b created
secret/kbs-auth-public-key-d64dkt2566 created
secret/keys-gmf2g75547 created
service/kbs created
deployment.apps/kbs created
Warning: annotation "kubernetes.io/ingress.class" is deprecated, please use 'spec.ingressClassName' instead
ingress.networking.k8s.io/kbs created
/tmp/kbs ~/go/src/github.com/kata-containers/kata-containers/tests
~/go/src/github.com/kata-containers/kata-containers/tests
pod/kbs-5699d4bc9f-89tb8 condition met
::endgroup::
::group::Check the service healthy
pod/kbs-checker-300525 created
pod "kbs-checker-300525" deleted
KBS service respond to requests
::endgroup::
::group::Check the kbs service is exposed
Trying to connect at kbs.77b2a8bfe3c8468385d0.eastus.aksapp.io. Timeout=350
KBS service respond to requests at kbs.77b2a8bfe3c8468385d0.eastus.aksapp.io
::endgroup::
``` 

Deleted as:
```shell
$ ./integration/kubernetes/gha-run.sh delete-coco-kbs
/tmp/kbs ~/go/src/github.com/kata-containers/kata-containers/tests
namespace "coco-tenant" deleted
configmap "kbs-config-k5gm89bk2b" deleted
secret "kbs-auth-public-key-d64dkt2566" deleted
secret "keys-gmf2g75547" deleted
service "kbs" deleted
deployment.apps "kbs" deleted
ingress.networking.k8s.io "kbs" deleted
~/go/src/github.com/kata-containers/kata-containers/tests
```

wainersm commented:
I'm working on more helper functions to be sent in another PR, to do operations like set a resource and policies to KBS. Initially I'm trying to use `curl` and avoid to build and install the `kbs_client` app. Ok, I'm stuck on the authentication part of the problem. Any expert watching this PR that could help?

Following is my set_resource function. To authenticate KBS, I build a JWT bearer token signed with the KBS' private key. I keep getting `Authentication failed: Requester is not an authorized user: token verification failed` error.
 
```
kbs_cli_set_resource() {
	local repository="${1:-}"
	local type="${2:-}"
	local tag="${3:-}"
	local data="${4:-)}"
	local host
	local port

	host=$(kbs_k8s_svc_host)
	port=$(kbs_k8s_svc_port)

	local url="http://${host}:${port}/kbs/v0/resource"
	[ -n "$repository" ] && url+="/$repository"
	url+="/$type"
	url+="/$tag"

	jwt_header="$(echo -n '{"alg":"EdDSA","typ":"JWT"}' | basenc --base64url)"
	# Set to expire in 4 hours from now
	local exp=$(($(date +%s)+(4*60*60)))
	jwt_payload="$(echo -n "{\"exp\":\"$exp\"}" | basenc --base64url)"
	local key="/tmp/kbs/kbs/config/kubernetes/base/kbs.key"

	echo "${jwt_header}.${jwt_payload}" > message.bin

	sign=$(openssl pkeyutl -sign -inkey "$key" -in message.bin -rawin | openssl enc -base64 |  tr -d '\n=' | tr -- '+/' '-_')
	token="${jwt_header}.${jwt_payload}.${sign}"

	curl -k -vvv -X POST "$url" -i \
		-H 'Content-Type: application/octet-stream' \
		-H "Authorization: Bearer $token" \
		-d "$data"
}
```

To be fair, I didn't test the KBS setup is well-configured. Probably I will give the kbs_client a try to ensure the setup and private key are correct.

wainersm commented:
> I'm working on more helper functions to be sent in another PR, to do operations like set a resource and policies to KBS. Initially I'm trying to use `curl` and avoid to build and install the `kbs_client` app. Ok, I'm stuck on the authentication part of the problem. Any expert watching this PR that could help?
> 
> Following is my set_resource function. To authenticate KBS, I build a JWT bearer token signed with the KBS' private key. I keep getting `Authentication failed: Requester is not an authorized user: token verification failed` error.
> 
> ```
> kbs_cli_set_resource() {
> 	local repository="${1:-}"
> 	local type="${2:-}"
> 	local tag="${3:-}"
> 	local data="${4:-)}"
> 	local host
> 	local port
> 
> 	host=$(kbs_k8s_svc_host)
> 	port=$(kbs_k8s_svc_port)
> 
> 	local url="http://${host}:${port}/kbs/v0/resource"
> 	[ -n "$repository" ] && url+="/$repository"
> 	url+="/$type"
> 	url+="/$tag"
> 
> 	jwt_header="$(echo -n '{"alg":"EdDSA","typ":"JWT"}' | basenc --base64url)"
> 	# Set to expire in 4 hours from now
> 	local exp=$(($(date +%s)+(4*60*60)))
> 	jwt_payload="$(echo -n "{\"exp\":\"$exp\"}" | basenc --base64url)"
> 	local key="/tmp/kbs/kbs/config/kubernetes/base/kbs.key"
> 
> 	echo "${jwt_header}.${jwt_payload}" > message.bin
> 
> 	sign=$(openssl pkeyutl -sign -inkey "$key" -in message.bin -rawin | openssl enc -base64 |  tr -d '\n=' | tr -- '+/' '-_')
> 	token="${jwt_header}.${jwt_payload}.${sign}"
> 
> 	curl -k -vvv -X POST "$url" -i \
> 		-H 'Content-Type: application/octet-stream' \
> 		-H "Authorization: Bearer $token" \
> 		-d "$data"
> }
> ```
> 
> To be fair, I didn't test the KBS setup is well-configured. Probably I will give the kbs_client a try to ensure the setup and private key are correct.

With kbs_client it does work, so my problem is really the JWT token:
```
[wmoschet@wainer-laptop kbs]$ ../target/release/kbs-client --url http://kbs.XXXXXXXXXXXXXX.eastu
s.aksapp.io config --auth-private-key /tmp/kbs/kbs/config/kubernetes/base/kbs.key set-resource --path 
default/aa/key --resource-file resource.file 
Set resource success 
 resource: a2V5X2FiY2QK
```

ldoktor commented on tests/gha-run-k8s-common.sh:
```diff
@@ -27,11 +27,15 @@ function _print_instance_type() {
 
+# Create the name string out of variables like $GH_PR_NUMBER (pull request
+# number) and the commit SHA. However, it can be overwritten by exporting
+# $AKS_NAME.
```
The commit message as well as the docstring is quite cryptic. How about simply: `Prints the cluster name set by "AKS_NAME" or generated out of runtime metadata`?

ldoktor commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -0,0 +1,111 @@
+	popd
+
+	if ! kubectl -n "$KBS_NS" wait --for=condition=Ready --timeout=2m \
+		-l app=kbs pod; then
```
I got `error: no matching resources found` here as the pod was not yet defined when we reached this. We should perhaps use `waitfor` for the pod definition first to avoid random failures in this section

wainersm replied:
This isn't actually odd, you've found a bug. The kbs pod is created out of a daemonset (ds), so it is perfectly fine it not exist in case the ds couldn't schedule pods. I will switch to the `waitfor` as you suggested. Thanks @ldoktor !

ldoktor commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -0,0 +1,111 @@
+	local pod=kbs-checker-$$
+	kubectl run "$pod" --image=busybox --restart=Never -- \
+		sh -c "wget -O- --timeout=5 \"${kbs_ip}:${kbs_port}\" || true"
+	if ! waitForProcess "60" "10" "kubectl logs \"$pod\" 2>/dev/null | grep -q \"404 Not Found\""; then
```
It would be nice to put a comment here explaining why are we looking for 404. It took me a while to understand this probe and I don't want to re-think this in few months.

Xynnn007 replied:
@wainersm One selection to cover @ldoktor 's issue (I have never tried, but seems similar with the current code inside this PR) is to use [a documented API of KBS](https://github.com/confidential-containers/kbs/blob/main/kbs/docs/kbs.yaml#L16-L46) and get an expected error. This would also makes it easier to edit comments.

wainersm commented:
I've just opened the PR #9116 that contains the first commit (45cb0b4dd671fd0b55576f292962e7a60c05b08a) of this PR. As it should be, first we merge the changes to the github workflow (in PR #9116) then we are allowed to run CI on this PR.

Ah, thanks for you review @ldoktor ! I should be addressing your comments on the next update of this PR.

wainersm commented:
Updates:

 - bumped to KBS version to commit 18c8ee378c6d83446ee635a702d5dee389028d8f so to pick https://github.com/confidential-containers/kbs/pull/324
 - using quay.io/prometheus/busybox insted of busybox to avoid hitting pull limits on bare-metal machines
 - Fixed the service health checker as @ldoktor suggested. Using `waitForProcess "60" "10" "kubectl logs \"$pod\" 2>/dev/null | grep -q \"404 Not Found\""`
 - re-phrased the "cryptic docstring" (@ldoktor )

ldoktor reviewed (approved):
Thanks, obviously the first commit needs to be modified according the https://github.com/kata-containers/kata-containers/pull/9116 but apart from that I like this.

Just for the record I have not tested it on AKS but I have tested it with kcli, the required changes are (I plan to send it as a follow-up when this is merged, adding it here to not to forget it and to simplify the life of reviewers without AKS ;-) ):

```patch
commit 147c064a8521abb504db82301efa8be8141d0372 (HEAD)
Author: Lukáš Doktor <ldoktor@redhat.com>
Date:   Mon Feb 26 11:14:09 2024 +0100

    tests/k8s: add nodeport ingress for KBS service
    
    The KBS service needs to be accessible from the outside, for kcli or
    clusters with public IP address the simplest way is to use nodeport
    service.
    
    This requires to have the KBS app deployed so the ingress deployment had
    to be moved at the end just before we check the ingress availability.
    
    Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

diff --git a/tests/integration/kubernetes/confidential_kbs.sh b/tests/integration/kubernetes/confidential_kbs.sh
index 173d66f46dd..54f0f31d684 100644
--- a/tests/integration/kubernetes/confidential_kbs.sh
+++ b/tests/integration/kubernetes/confidential_kbs.sh
@@ -81,8 +81,6 @@ function kbs_k8s_deploy() {
        popd
        echo "::endgroup::"
 
-       [ -n "$ingress" ] && _handle_ingress "$ingress"
-
        echo "::group::Deploy the KBS"
        ./deploy-kbs.sh
        popd
@@ -126,6 +124,7 @@ function kbs_k8s_deploy() {
        echo "::endgroup::"
 
        if [ -n "$ingress" ]; then
+               _handle_ingress "$ingress"
                echo "::group::Check the kbs service is exposed"
                svc_host=$(kbs_k8s_svc_host)
                if [ -z "$svc_host" ]; then
@@ -147,13 +146,21 @@ function kbs_k8s_deploy() {
        fi
 }
 
-# Return the kbs service host name in case ingress is configured
-# otherwise the cluster IP.
+# Prints the kbs address. In case of an ingress it returns the
+# host name, for nodeport a cluster ip:port, otherwise the internal
+# cluster IP.
 #
 kbs_k8s_svc_host() {
        if kubectl get ingress -n "$KBS_NS" | grep -q kbs; then
                kubectl get ingress kbs -n "$KBS_NS" \
                        -o jsonpath='{.spec.rules[0].host}' 2>/dev/null
+       elif kubectl get svc kbs-nodeport -n "$KBS_NS" &>/dev/null; then
+               local host
+               local port
+               host=$(kubectl get nodes -o jsonpath='{.items[0].status.addresses[?(@.type=="ExternalIP")].address}' -n "$KBS_NS")
+               [ -z "$host"] && host=$(oc get nodes -o jsonpath='{.items[0].status.addresses[?(@.type=="InternalIP")].address}' -n "$KBS_NS")
+               port=$(kubectl get -o jsonpath='{.spec.ports[0].nodePort}' svc kbs-nodeport -n "$KBS_NS")
+               echo "$host:$port"
        else
                kubectl get svc kbs -n "$KBS_NS" \
                        -o jsonpath='{.spec.clusterIP}' 2>/dev/null
@@ -209,4 +216,14 @@ _handle_ingress_aks() {
 
        kustomize edit add resource ingress.yaml
        popd
-}
\ No newline at end of file
+}
+
+# Implements the ingress handler for servernode
+# this is useful on kcli or anywhere where cluster IPs are accessible
+# from the testing machines.
+#
+_handle_ingress_nodeport() {
+       local kbs_port
+       kbs_port=$(kubectl get -o jsonpath='{.spec.ports[0].port}' svc "$KBS_SVC_NAME" -n "$KBS_NS" 2>/dev/null)
+       kubectl expose deployment kbs --name=kbs-nodeport --port="$kbs_port" --target-port="$kbs_port" --selector='app=kbs' --type=NodePort -n "$KBS_NS"
+}
```

wainersm commented:
The pre-req PR https://github.com/kata-containers/kata-containers/pull/9116 was just merge. This means the  https://github.com/kata-containers/kata-containers/blob/main/.github/workflows/run-k8s-tests-on-aks.yaml workflow will get the KBS installed if `vmm==qemu` and `host_os==ubuntu`, i.e., it will run on a non-TEE job first. More precisely, the `Deploy CoCo KBS` step (https://github.com/kata-containers/kata-containers/blob/main/.github/workflows/run-k8s-tests-on-aks.yaml#L111) should execute.

I'll add the `ok-to-test` label. Fingers crossed!

wainersm commented:
I've already a fix to `Static checks / static-checks (make static-checks) (pull_request)`. Gonna sent later.

wainersm commented:
ok, I've broken the step to deploy kata: /home/runner/work/kata-containers/kata-containers/tests/integration/kubernetes/../../gha-run-k8s-common.sh: line 37: AKS_NAME: unbound variable

Fixing it...

wainersm commented:
KBS deployment failed (e.g. https://github.com/kata-containers/kata-containers/actions/runs/8087248754/job/22101339904?pr=9065):
```
Clone the kbs sources
/tmp/kbs/kbs/config/kubernetes /tmp/kbs ~/work/kata-containers/kata-containers
Update the kbs container image
WARNING: The behavior of this command has been altered by the following extension: aks-preview
ERROR: (ResourceGroupNotFound) Resource group 'kataCI-k8s-9065-18c8ee378c6d-qemu-ubuntu-amd64-s' could not be found.
Code: ResourceGroupNotFound
Message: Resource group 'kataCI-k8s-9065-18c8ee378c6d-qemu-ubuntu-amd64-s' could not be found.
Error: Process completed with exit code 3.
```

It seems to be in this call [az aks show -g "$rg" -n "$cluster_name" --query "$q" | tr -d \"](https://github.com/kata-containers/kata-containers/pull/9065/commits/ca4fcdcc40cce90d09cc3dac58f59bf83ad597fb#diff-ed957d9b9df3f78e31286a82bbf4df7babd2f8ef55b0be3b49e66651f69b4ef4R162)

wainersm commented:
> KBS deployment failed (e.g. https://github.com/kata-containers/kata-containers/actions/runs/8087248754/job/22101339904?pr=9065):
> 
> ```
> Clone the kbs sources
> /tmp/kbs/kbs/config/kubernetes /tmp/kbs ~/work/kata-containers/kata-containers
> Update the kbs container image
> WARNING: The behavior of this command has been altered by the following extension: aks-preview
> ERROR: (ResourceGroupNotFound) Resource group 'kataCI-k8s-9065-18c8ee378c6d-qemu-ubuntu-amd64-s' could not be found.
> Code: ResourceGroupNotFound
> Message: Resource group 'kataCI-k8s-9065-18c8ee378c6d-qemu-ubuntu-amd64-s' could not be found.
> Error: Process completed with exit code 3.
> ```
> 
> It seems to be in this call [az aks show -g "$rg" -n "$cluster_name" --query "$q" | tr -d "](https://github.com/kata-containers/kata-containers/pull/9065/commits/ca4fcdcc40cce90d09cc3dac58f59bf83ad597fb#diff-ed957d9b9df3f78e31286a82bbf4df7babd2f8ef55b0be3b49e66651f69b4ef4R162)

I just pushed a fix for this problem. @sprt realized it is getting the wrong commit on the resources group (`$rg`) name. The commit I pushed should be squashed before merge and I explain the problem/solution inline. I hope it solve the problem, I didn't try :)

wainersm commented:
> > KBS deployment failed (e.g. https://github.com/kata-containers/kata-containers/actions/runs/8087248754/job/22101339904?pr=9065):
> > ```
> > Clone the kbs sources
> > /tmp/kbs/kbs/config/kubernetes /tmp/kbs ~/work/kata-containers/kata-containers
> > Update the kbs container image
> > WARNING: The behavior of this command has been altered by the following extension: aks-preview
> > ERROR: (ResourceGroupNotFound) Resource group 'kataCI-k8s-9065-18c8ee378c6d-qemu-ubuntu-amd64-s' could not be found.
> > Code: ResourceGroupNotFound
> > Message: Resource group 'kataCI-k8s-9065-18c8ee378c6d-qemu-ubuntu-amd64-s' could not be found.
> > Error: Process completed with exit code 3.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It seems to be in this call [az aks show -g "$rg" -n "$cluster_name" --query "$q" | tr -d "](https://github.com/kata-containers/kata-containers/pull/9065/commits/ca4fcdcc40cce90d09cc3dac58f59bf83ad597fb#diff-ed957d9b9df3f78e31286a82bbf4df7babd2f8ef55b0be3b49e66651f69b4ef4R162)
> 
> I just pushed a fix for this problem. @sprt realized it is getting the wrong commit on the resources group (`$rg`) name. The commit I pushed should be squashed before merge and I explain the problem/solution inline. I hope it solve the problem, I didn't try :)

Make another fix on the same commit. I wanted to export AKS_NAME in case it is not, so using `-z` instead `-n` in the conditional.

wainersm commented:
> > > KBS deployment failed (e.g. https://github.com/kata-containers/kata-containers/actions/runs/8087248754/job/22101339904?pr=9065):
> > > ```
> > > Clone the kbs sources
> > > /tmp/kbs/kbs/config/kubernetes /tmp/kbs ~/work/kata-containers/kata-containers
> > > Update the kbs container image
> > > WARNING: The behavior of this command has been altered by the following extension: aks-preview
> > > ERROR: (ResourceGroupNotFound) Resource group 'kataCI-k8s-9065-18c8ee378c6d-qemu-ubuntu-amd64-s' could not be found.
> > > Code: ResourceGroupNotFound
> > > Message: Resource group 'kataCI-k8s-9065-18c8ee378c6d-qemu-ubuntu-amd64-s' could not be found.
> > > Error: Process completed with exit code 3.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > It seems to be in this call [az aks show -g "$rg" -n "$cluster_name" --query "$q" | tr -d "](https://github.com/kata-containers/kata-containers/pull/9065/commits/ca4fcdcc40cce90d09cc3dac58f59bf83ad597fb#diff-ed957d9b9df3f78e31286a82bbf4df7babd2f8ef55b0be3b49e66651f69b4ef4R162)
> > 
> > 
> > I just pushed a fix for this problem. @sprt realized it is getting the wrong commit on the resources group (`$rg`) name. The commit I pushed should be squashed before merge and I explain the problem/solution inline. I hope it solve the problem, I didn't try :)
> 
> Made another fix on the same commit. I wanted to export AKS_NAME in case it is not, so using `-z` instead `-n` in the conditional.

This problem is fixed! However, I will leave the commit 02b4981379655d62d54213a4649bd86cf5af5e9a around for a while in case someone has any suggestions. To be honest, I'm not totally happy with the solution but at same time I think it is good enough for this first interaction.

wainersm commented:
ok, KBS was installed on two jobs; one passed and the other failed:

 * PASSED -> [kata-containers-ci-on-push / run-k8s-tests-on-aks / run-k8s-tests (ubuntu, qemu, small)](https://github.com/kata-containers/kata-containers/actions/runs/8103661265/job/22149331492?pr=9065#logs)
 * FAILED -> [kata-containers-ci-on-push / run-k8s-tests-on-aks / run-k8s-tests (ubuntu, qemu, normal)](https://github.com/kata-containers/kata-containers/actions/runs/8103661265/job/22149331943?pr=9065#logs)

The failed one, it seemed to deploy the ingress just fine but it failed while checking the service was exposed. Actually the checking didn't hit a timeout, but rather the step `  Error: The action has timed out.` . 

The step timeout is 5 min (https://github.com/kata-containers/kata-containers/blob/main/.github/workflows/run-k8s-tests-on-aks.yaml#L113) which is the same value for the service checking. Thus I should increase the timeout of the step. Maybe set to 10min.

I will run the jobs a couple of time more to see if any other intermittent error pop up.

wainersm commented:
> ok, KBS was installed on two jobs; one passed and the other failed:
> 
>     * PASSED -> [kata-containers-ci-on-push / run-k8s-tests-on-aks / run-k8s-tests (ubuntu, qemu, small)](https://github.com/kata-containers/kata-containers/actions/runs/8103661265/job/22149331492?pr=9065#logs)
> 
>     * FAILED -> [kata-containers-ci-on-push / run-k8s-tests-on-aks / run-k8s-tests (ubuntu, qemu, normal)](https://github.com/kata-containers/kata-containers/actions/runs/8103661265/job/22149331943?pr=9065#logs)
> 
> 
> The failed one, it seemed to deploy the ingress just fine but it failed while checking the service was exposed. Actually the checking didn't hit a timeout, but rather the step ` Error: The action has timed out.` .
> 
> The step timeout is 5 min (https://github.com/kata-containers/kata-containers/blob/main/.github/workflows/run-k8s-tests-on-aks.yaml#L113) which is the same value for the service checking. Thus I should increase the timeout of the step. Maybe set to 10min.
> 
> I will run the jobs a couple of time more to see if any other intermittent error pop up.

Here is the PR that increases the timeout: https://github.com/kata-containers/kata-containers/pull/9200 (please reviews!)

wainersm commented:
Rebased to main as https://github.com/kata-containers/kata-containers/pull/9200 was merged. Let's see if it deploys KBS consistently

wainersm commented:
/test

wainersm commented:
hi all,

It seems to consistently deploy KBS. I've also tested the follow on https://github.com/kata-containers/kata-containers/pull/9114 successfully. I know that there are a few comments still not addressed but I'd like to propose them being on another follow on PR; this way we merge this PR which seems good enough.

ChengyuZhu6 reviewed (approved):
LGTM thanks @wainersm !
